src/dataframe/dataframe.jl

```diff
@@ -1194,14 +1204,49 @@ function Base.append!(df1::DataFrame, df2::AbstractDataFrame)
     end
     ncol(df2) == 0 && return df1
 
-    _names(df1) == _names(df2) || throw(ArgumentError("Column names do not match"))
+    if cols == :orderequal && _names(df1) != _names(df2)
+        if ncol(df1) != ncol(df2)
+            wrongnames = symdiff(Set(_names(df1)), Set(_names(df2)))
+            if length(wrongnames) > 0
+                mismatchmsg = " Column names :" *
+                              join(wrongnames, ", :", " and :") *
+                              "were found in only one of the passed data frames"
+            else
+                mismatchmsg = ""
+            end
+            throw(ArgumentError("The number of columns in the two data frames does " *
+                                "not match and passed `cols` value is `:orderequal`.") *
+                                mismatchmsg)
+        end
+        mismatches = findall(_names(df1) .!= _names(df2))
+        throw(ArgumentError("Columns number " *
+                            join(mismatches, ", ", " and ") *
+                            " do not have the same names in both passed data frames" *
+                            "and `cols=:orderequal`"))
+    end
+    if cols == :setequal
+        ns1 = Set(_names(df1))
```

Comment:
Here `ns1` and `ns2` don't need to be stored in variables either.

Also are you sure calling `Set` is really useful here and above? We don't care about performance in this error path anyway.

Reply:
We can change this after #2022 is merged as then we will be sure that `df2` does not contain duplicates. Now we had to use `Set`, because:
```
julia> symdiff([1,2,3], [1,1,2,2,3,3])
3-element Array{Int64,1}:
 1
 2
 3
```